File: python/model/compressed_mask.py
import zenoh
from edgefirst.schemas.edgefirst_msgs import Mask
import rerun as rr
from argparse import ArgumentParser
import sys
import numpy as np
import zstd

def main():
    args = ArgumentParser(description="EdgeFirst Samples - Compressed Mask")
    args.add_argument('-r', '--remote', type=str, default=None,
                      help="Connect to a Zenoh router rather than local.")
    rr.script_add_args(args)
    args = args.parse_args()

    rr.script_setup(args, "mask")

    # Create the default Zenoh configuration and if the connect argument is
    # provided set the mode to client and add the target to the endpoints.
    config = zenoh.Config()
    if args.remote is not None:
        config.insert_json5("mode", "'client'")
        config.insert_json5("connect", '{"endpoints": ["%s"]}' % args.remote)
    session = zenoh.open(config)

    # Create a subscriber for "rt/camera/mask_compressed"
    subscriber = session.declare_subscriber('rt/model/mask_compressed')

    while True:
        msg = subscriber.recv()
        mask = Mask.deserialize(msg.payload.to_bytes())
        decoded_array = zstd.decompress(bytes(mask.mask))
        np_arr = np.frombuffer(decoded_array, np.uint8)
        np_arr = np.reshape(np_arr, [mask.height, mask.width, -1])
        np_arr = np.argmax(np_arr, axis=2)
        rr.log("/", rr.AnnotationContext([(0, "background", (0,0,0)), (1, "person", (0,255,0))]))
        rr.log("mask", rr.SegmentationImage(np_arr))


if __name__ == "__main__":    
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)

File: rust/model/compressed_mask.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright © 2025 Au-Zone Technologies. All Rights Reserved.

use clap::Parser as _;
use edgefirst_samples::Args;
use edgefirst_schemas::edgefirst_msgs::Mask;
use ndarray::{Array, Array2};
use rerun::{AnnotationContext, SegmentationImage};
use std::error::Error;
use std::io::Cursor;
use zstd::stream::decode_all;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args = Args::parse();
    let session = zenoh::open(args.clone()).await.unwrap();

    // Create a subscriber for "rt/model/mask"
    let subscriber = session
        .declare_subscriber("rt/model/mask_compressed")
        .await
        .unwrap();

    // Create Rerun logger using the provided parameters
    let (rr, _serve_guard) = args.rerun.init("model-compressed_mask")?;

    while let Ok(msg) = subscriber.recv() {
        let mask: Mask = cdr::deserialize(&msg.payload().to_bytes())?;
        let decompressed_bytes = decode_all(Cursor::new(&mask.mask))?;

        let h = mask.height as usize;
        let w = mask.width as usize;
        let total_len = mask.mask.len() as u32;
        let c = (total_len / (h as u32 * w as u32)) as usize;

        let arr3 = Array::from_shape_vec([h, w, c], decompressed_bytes.clone())?;

        // Compute argmax along the last axis (class channel)
        let array2: Array2<u8> = arr3.map_axis(ndarray::Axis(2), |class_scores| {
            class_scores
                .iter()
                .enumerate()
                .max_by_key(|(_, val)| *val)
                .map(|(idx, _)| idx as u8)
                .unwrap_or(0)
        });

        // Log annotation context
        rr.log(
            "/",
            &AnnotationContext::new([
                (0, "background", rerun::Rgba32::from_rgb(0, 0, 0)),
                (1, "person", rerun::Rgba32::from_rgb(0, 255, 0)),
            ]),
        )?;

        // Log segmentation mask
        rr.log("mask", &SegmentationImage::try_from(array2)?)?;
    }

    Ok(())
}
